### PR TITLE
Add support for flashing over USB and UART

### PR DESCRIPTION
**Pull request by KevinOConnor**
This PR extends CanBoot so that it is possible to flash over USB and UART (along with the existing CANbus support).

This PR is a bit large.  I wanted to publish what I have done so far in order to get feedback on the approach.

To implement this support I reworked much of the CanBoot mcu code so that it has a more similar layout to the Klipper code.  This makes it easier to import the Klipper low-level device drivers for USB and UART.

Ultimately, I was able to pull in usbfs.c, serial.c, stm32f0_serial.c, usb_cdc.c, etc. with almost no change to those files.  Similarly, I think it would be possible to pull in other architectures (eg, samd, rp2040) with little changes.

To achieve this, I ended up adding a bunch of wrapper functions (eg, `sched_wake_task()`) to the code.  I've also reworked much of the command parsing.  I'm "a bit torn" on these changes as it does make the CanBoot code more complicated.  It may be possible to rework the upstream Klipper command interface a little so that it's easier to export the low-level functionality.

The wrapper functions don't noticeably increase the binary size.  In general, gcc does a good job of inlining and pruning unused code.  On my build, compiling for can on stm32f103 (with led and button support) used to be 3604 bytes and is now 3700 bytes.  I think the extra 96 bytes is mostly because the previous code invoked the command handlers directly from irq context, while the new code does that work in a task (as Klipper does).

I tested this on an stm32f103 with serial, usb, and canbus.  There is a new `-d` option to flash_can.py to support flashing over serial.  The flash_can.py support requires pyserial to be installed (in order to flash over usb or serial).  The flash_can.py support is primitive at this point - I'm sure it could be further enhanced.

Thoughts?
-Kevin

### CODE REVIEW

**Comment by Arksine** [comment]
Hi Kevin, I left a few comments, however overall it looks good to me.  I don't have an issue with adapting the wrappers, I can see value in adding task management similar to Klipper's.  Being able to pull in the platform specific code without significant modification is a huge plus.

I think the addition to `flash_can.py` looks good.  I somewhat wonder if it would be possible to add usb devices to the `query` method (returns device location instead of a node id), however that is something I'd be willing to look at later on.

File: src/command.c [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+{
+    out[1] = cpu_to_le32(acked_cmd);
+    command_respond(out, RESPONSE_ACK, out_len);
+}
+
+void
+command_respond_command_error(void)
+{
+    uint32_t out[2];
+    command_respond(out, RESPONSE_COMMAND_ERROR, 0);
```
[the remark this answers is not in the document]
Comment (KevinOConnor) [reply src/command.c]: Good catch.  I fixed that error and rebased this PR.

I'm not sure what you mean by calculating in `command_respond()` (the ARRAY_SIZE macro doesn't work on pointers).

-Kevin

**Comment by KevinOConnor** [comment]
I guess one could implement `-q` for usb.  However, it's pretty easy to run `ls -l /dev/serial/by-id/usb-CanBoot*`.

The flash_can.py script could also implement the 1200 baud trick to enter the bootloader from Klipper.  (The `enter_bootloader()` from Klipper's flash_usb.py script.)  I can add that as a follow up.

-Kevin

File: scripts/flash_can.py [hunk 2]
```diff
@@ -526,23 +586,33 @@ def main():
     intf = args.interface
     fpath = pathlib.Path(args.firmware).expanduser().resolve()
     loop = asyncio.get_event_loop()
+    iscan = args.uuid is not None
```
[the remark this answers is not in the document]
Comment (KevinOConnor) [reply scripts/flash_can.py]: Good point.  Fixed and rebased also.

-Kevin

File: src/command.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (Arksine) [reply src/command.c]: Yeah, I realized that about a minute after I posted it.  Unfortunately editing the reply doesn't change the email notification.

**Comment by Arksine** [comment]
Thanks Kevin, everything looks good. I tested CAN and and USB on a bluepill.  USB flashing is fast, I like it!   This is a pretty exciting addition since it opens up the door to support nearly any ARM device that runs Klipper.

**Pull request merged**